In [1]:
import pandas as pd
import numpy as np
import re
import glob

In [3]:
data1 = pd.read_csv('../../latest/original.csv')
data2 = pd.read_csv('../../latest/sudharma.csv')
data3 = pd.read_csv('../../latest/navavani.csv')

In [4]:
print(data1.isna().sum())
print(data2.isna().sum())
print(data3.isna().sum())

text       0
summary    0
dtype: int64
text       38
summary     0
dtype: int64
text       0
summary    0
dtype: int64


In [5]:
print(data1.duplicated().sum())
print(data2.duplicated().sum())
print(data3.duplicated().sum())

3
0
0


In [6]:
data1.shape

(6515, 2)

In [7]:
data2.shape

(6002, 2)

In [8]:
data3.shape

(1224, 2)

In [9]:
len(data1)+len(data2)+len(data3)

13741

In [16]:
data1[data1['summary'] == 'संक्षिप्तवार्ताः']

text           summary
2     जम्मूकश्मीर सुरक्षाबलैः तैयब्बा क्रूरकर्मनिरतस...  संक्षिप्तवार्ताः
192   केन्द्रीयमार्गनिर्माणविभागेन\nविश्वविक्रमः स्थ...  संक्षिप्तवार्ताः
636      कोशी-मिथीलानगरयोः मध्ये रेलसेवा पुनः प्रारब्धा  संक्षिप्तवार्ताः
669   इराकदेशस्य प्रधानमन्त्री मुस्तफा अल-कदीमी ईरान...  संक्षिप्तवार्ताः
920              ज्ञानवापी विवादोपरि भागवतस्य वक्तव्यम्  संक्षिप्तवार्ताः
937              हिन्दूराष्ट्रे सर्वधर्मसमन्वयः आवश्यकः  संक्षिप्तवार्ताः
1052  आरसीपी सिंहस्य यथायोग्यं मान्यं पूर्वमेव प्रदत...  संक्षिप्तवार्ताः
1617         शपथग्रहणसमारोहस्थले प्राक्तनमन्त्री कुपितः  संक्षिप्तवार्ताः
1716  प्रधानमन्त्रिणा नरेन्द्रमोदिना सर्वेषु विभागेष...  संक्षिप्तवार्ताः
1911             श्रीलंकायां तमिलनागरिकाः उपकृताः स्युः  संक्षिप्तवार्ताः
2515             विवादितप्रश्नमाधृत्य छात्रेषु असन्तोषः  संक्षिप्तवार्ताः
3100     हरियाणाराज्यं खलिस्तानी-प्रदेशरुपेण स्वीकरणीयः  संक्षिप्तवार्ताः
3848  सिंहदेव-बंसलयो नेतृत्वे राज्यसभायैः निर्वाचनं ...  संक्षिप्तवार्ताः
3922  प्रवर्त्तननिदेशालयः कदाचित् स्मारकीस्थलोपरि कट...  संक्षिप्तवार्ताः
4180  पूर्व-मध्य रेल विभागेन बिहार प्रदेशात दूरगामिन...  संक्षिप्तवार्ताः
4373                 महिलाकर्मिणां कृते समयः निर्धारितः  संक्षिप्तवार्ताः
4452  उद्धव ठाकरे मुख्यमंत्रिणं ‘एकनाथ शिंदे’ शिवसेन...  संक्षिप्तवार्ताः
4926                   अजय कोठीयालः आपसंगठनं त्यक्तवान्  संक्षिप्तवार्ताः
5970  वैश्विक महामारी-संक्रमणे चीने वुहानस्य हुआनन-स...  संक्षिप्तवार्ताः
6231  प्रधानमन्त्री नरेन्द्र मोदी सविशेष उक्तवान् यत...  संक्षिप्तवार्ताः
6366               जम्मूक्षेत्रे हिजबूलसदस्यः अधिगृहीतः  संक्षिप्तवार्ताः

In [10]:
data1.head()

text  \
0  व्याकरणात् संस्कृतभाषायां संस्कार आयाति। द्विस...   
1  मारिया विश्वस्तरीय१८स्थाने अस्ति। मारिया “यूएस...   
2  जम्मूकश्मीर सुरक्षाबलैः तैयब्बा क्रूरकर्मनिरतस...   
3  जनुवरि प्रथमदिनाङ्कतः आरभ्य पलास्तिक कूप्यां व...   
4  गीतस्वरैश्चित्तहरैस्तदीयैः प्रत्यक्षमेनां शृणु...   

                                             summary  
0                                            संस्कृत  
1  प्रक्षिप्त-कन्दुकक्रीडायै मारिया अग्रिम क्रीडा...  
2                                   संक्षिप्तवार्ताः  
3         पलास्तिककूपीजलं त्यजामः, आरोग्यं संरक्षामः  
4                               गानदेवी-लता मंगेस्कर

In [11]:
data2.head()

text  \
0  बीजिङ्ग - भारते मार्चमासस्य कार्यं चिकीर्षति ।...   
1  इन्दोर दिस्त असतेयः विश न मास्ति मध्यदेशस्य कु...   
2  डा. विश्वास म भरद्वाज - यवक्रीत - रैभ्यादीनां ...   
3  सेवाभ निर्माणीयम् स्वामिनोऽनुवन्। भानुवासरे आय...   
4  इति एकस्मिन्ार्य कार्यक्र अमर एकदा प्रधान श्री...   

                                             summary  
0           चीनाविदेशाङ्गमन्त्री जी-२० शृङ्गसभायाम्   
1                                    विश्वेभ्यः हार   
2                   महाभारतान्तर्गताः काश्चन उपकथाः   
3                       सुवर्णपुष्पाभिवृष्टिः सेवाभ   
4  विपक्षा एवं हासयन्ति स्म ! सुकले प्रधानम प्रधानम

In [20]:
data = pd.concat([data1, data2, data3], ignore_index = True)

In [21]:
data.shape

(13741, 2)

In [22]:
print(data.isna().sum())
print(data.duplicated().sum())

text       38
summary     0
dtype: int64
3


In [23]:
data.drop_duplicates(inplace = True)

In [24]:
data.shape

(13738, 2)

In [25]:
data.dropna(inplace = True)

In [26]:
data.shape

(13700, 2)

In [27]:
def basic_clean(df):
    """Basic Cleaning function
    
    Args:
        df (DataFrame): a dataframe column to clean
        
    Returns:
        df (DataFrame): cleaned dataframe
    """
    #df = df.apply(lambda x : re.sub("॥.+॥","।",x)) #removing endingwords of sentence eg. '॥ ३-२०॥'
    df = df.apply(lambda x : re.sub("[a-zA-Z_]","",x)) #removing english chars
    df = df.apply(lambda x: re.sub(" {2,}","",x)) #removing extra space between the words in a sentence
    df = df.apply(lambda x: re.sub(r'\([^()]*\)', '',x)) 
    df = df.apply(lambda x: re.sub("[./:]","",x))
    df = df.apply(lambda x: re.sub("(\n)+","\n",x)) #removing more than one next line character
    return df

def remove_char(cha, st): # helper function for selective_clean
    before, sep, after = st.partition(cha)
    if len(after)>0:
        return after
    return before

def selective_clean(cha, df):
    """Remove a specific character from the dataframe. Used for selective cleaning."""
    df = df.apply(lambda x: remove_char(cha, x)) # remove text before a specific character
    return df

In [28]:
data.head()

text  \
0  व्याकरणात् संस्कृतभाषायां संस्कार आयाति। द्विस...   
1  मारिया विश्वस्तरीय१८स्थाने अस्ति। मारिया “यूएस...   
2  जम्मूकश्मीर सुरक्षाबलैः तैयब्बा क्रूरकर्मनिरतस...   
3  जनुवरि प्रथमदिनाङ्कतः आरभ्य पलास्तिक कूप्यां व...   
4  गीतस्वरैश्चित्तहरैस्तदीयैः प्रत्यक्षमेनां शृणु...   

                                             summary  
0                                            संस्कृत  
1  प्रक्षिप्त-कन्दुकक्रीडायै मारिया अग्रिम क्रीडा...  
2                                   संक्षिप्तवार्ताः  
3         पलास्तिककूपीजलं त्यजामः, आरोग्यं संरक्षामः  
4                               गानदेवी-लता मंगेस्कर

In [29]:
import copy
ndata1 = copy.deepcopy(data)

In [30]:
ndata1['text'] = basic_clean(ndata1['text'])
ndata1['summary'] = basic_clean(ndata1['summary'])

In [31]:
ndata1.head()

text  \
0  व्याकरणात् संस्कृतभाषायां संस्कार आयाति। द्विस...   
1  मारिया विश्वस्तरीय१८स्थाने अस्ति। मारिया “यूएस...   
2  जम्मूकश्मीर सुरक्षाबलैः तैयब्बा क्रूरकर्मनिरतस...   
3  जनुवरि प्रथमदिनाङ्कतः आरभ्य पलास्तिक कूप्यां व...   
4  गीतस्वरैश्चित्तहरैस्तदीयैः प्रत्यक्षमेनां शृणु...   

                                             summary  
0                                            संस्कृत  
1  प्रक्षिप्त-कन्दुकक्रीडायै मारिया अग्रिम क्रीडा...  
2                                   संक्षिप्तवार्ताः  
3         पलास्तिककूपीजलं त्यजामः, आरोग्यं संरक्षामः  
4                               गानदेवी-लता मंगेस्कर

In [35]:
ndata1.loc[12493]

text                
summary    सुभाषितम्
Name: 12493, dtype: object

In [40]:
ndata1.duplicated(subset = ['text', 'summary']).sum()

41

In [41]:
df = ndata1.drop_duplicates()

In [42]:
df.isna().sum()

text       0
summary    0
dtype: int64

In [43]:
df.duplicated().sum()

0

In [44]:
df.isna().sum()

text       0
summary    0
dtype: int64

In [47]:
df.shape

(13659, 2)

In [70]:
duplicates_text = df[df['text'].duplicated()]

In [71]:
duplicates_summary = df[df['summary'].duplicated()]

In [74]:
total_indices = set(df[df['summary'].duplicated()].index).union(set(df[df['text'].duplicated()].index))

In [61]:
846+44-17

873

In [62]:
13659-873

12786

In [65]:
12786 - 1000 - 500

11286

In [75]:
clean_df = df.drop(index = total_indices)

In [84]:
clean_df.shape

(12786, 2)

In [86]:
clean_df_idx_reset = clean_df.reset_index(drop = True)

In [87]:
clean_df_idx_reset.to_csv('../for_BERT/uclean_data/pre_short_sentences_removal.csv', index = False)

<h1>FIRST SAVE----------------------------------------------------------------------------------------------</h1>

In [171]:
def find_indices(sub_df, thresh):
    """Find the indices of rows which have a text lesser than 'threshold' words.
    
    Args:
        sub_df (DataFrame): the dataframe to check
        thresh (int): no of words threshold
        
    Returns:
        idxs (set): a set of indices of rows which have words lesser than the thresh
    """
    idxs = set()
    for idx, row in sub_df.iterrows():
        # splits a sentence based on newline and space to check if the length is lesser than the threshold
        if len(re.split('\s+', row['text'])) <=thresh:
            idxs.add(idx)
    return idxs

In [172]:
x = find_indices(clean_df_idx_reset, 6)

In [173]:
len(x)

521

In [174]:
uclean_df = clean_df_idx_reset.drop(index = x).reset_index(drop = True)

In [175]:
uclean_df

text  \
0      व्याकरणात् संस्कृतभाषायां संस्कार आयाति। द्विस...   
1      मारिया विश्वस्तरीय१८स्थाने अस्ति। मारिया “यूएस...   
2      जम्मूकश्मीर सुरक्षाबलैः तैयब्बा क्रूरकर्मनिरतस...   
3      जनुवरि प्रथमदिनाङ्कतः आरभ्य पलास्तिक कूप्यां व...   
4      गीतस्वरैश्चित्तहरैस्तदीयैः प्रत्यक्षमेनां शृणु...   
...                                                  ...   
12260  नवदिल्ली – उपराष्ट्रपतित्वेन निर्वाचितः वेङ्कय...   
12261  नवदिल्ली – भारते पचनवायुकोशानां कृते दीयमानं व...   
12262  नवदिल्ली – विद्यालयेषु अष्टमकक्ष्यापर्यन्तं प्...   
12263  पियु चित्रा\nकोच्ची – विविश्वकुशलस्पर्धायां भा...   
12264  इस्लामबाद्\nपाकिस्थानस्य प्रधानमन्त्री नवास षर...   

                                                 summary  
0                                                संस्कृत  
1      प्रक्षिप्त-कन्दुकक्रीडायै मारिया अग्रिम क्रीडा...  
2                                       संक्षिप्तवार्ताः  
3             पलास्तिककूपीजलं त्यजामः, आरोग्यं संरक्षामः  
4                                   गानदेवी-लता मंगेस्कर  
...                                                  ...  
12260                  उपराष्ट्रपतेः शपथवचनं शुक्रवासरे।  
12261                   पचनवायोः वित्तानुकूल्यं स्थगयति।  
12262            विद्यालयेषु इतः परं असमर्थानामवतीर्णता।  
12263              पि यु चित्रा विश्वस्पर्धात बहिष्कृता।  
12264                          नवास् षरीफः स्थानभ्रष्टः।  

[12265 rows x 2 columns]

In [176]:
uclean_df.to_csv('../for_BERT/uclean_data/df_removed_6len_words.csv', index = False)

<h1>CLEAN SAVE--------------------------------------------------------------------------------------------</h1><h3> 12265 rows </h3> 

In [178]:
from sklearn.model_selection import train_test_split

In [179]:
y = uclean_df['summary']
X = uclean_df['text']

In [230]:
def calc_dims(X_shape, val_perc, test_perc):
    train_first = X_shape - val_perc
    train_val = 100 - (train_first*100/X_shape)
    val_first = val_perc - test_perc
    val_val = 100 - (val_first*100/val_perc)
    return train_val, val_val

In [285]:
def stats(df, label = None):
    print(f'{label} : {df.shape}')
    #print(f'Null: {df.isna().sum()}\nDuplicates: {df.duplicated().sum()}')
    for col in df.columns:
        print(f'Null in {col}: {df[col].isna().sum()}')
        print(f'Duplicates in {col}: {df.duplicated(subset = [col]).sum()}')

In [251]:
calc_dims(len(X), 1745, 245)

(14.227476559315122, 14.04011461318052)

In [252]:
Xtr, Xtest, ytr, ytest = train_test_split(X,y,test_size = .14227,shuffle = True)
Xval, Xtest, yval, ytest = train_test_split(Xtest, ytest, test_size = .14040, shuffle = True)

In [253]:
print(f'Train data shape : {Xtr.shape}, {ytr.shape}\nValidation data shape:{Xval.shape}, {yval.shape}\nTest data shape: {Xtest.shape}, {ytest.shape}')

Train data shape : (10520,), (10520,)
Validation data shape:(1500,), (1500,)
Test data shape: (245,), (245,)


In [254]:
print('Training Data Dimensions: ', Xtr.shape, ytr.shape)
print('Validation Data Dimensions: ', Xval.shape, yval.shape)
print('Testing Data Dimensions: ', Xtest.shape, ytest.shape)

Training Data Dimensions:  (10520,) (10520,)
Validation Data Dimensions:  (1500,) (1500,)
Testing Data Dimensions:  (245,) (245,)


In [287]:
stats(uclean_df, 'uclean_df')

uclean_df : (12265, 2)
Null in text: 0
Duplicates in text: 0
Null in summary: 0
Duplicates in summary: 0


In [280]:
final_train_data = pd.DataFrame({'text':Xtr, 'summary':ytr}).reset_index(drop = True)

In [281]:
final_test_data = pd.DataFrame({'text':Xtest, 'summary':ytest}).reset_index(drop = True)

In [282]:
final_val_data = pd.DataFrame({'text':Xval, 'summary':yval}).reset_index(drop = True)

In [290]:
stats(final_train_data, 'train')

train : (10520, 2)
Null in text: 0
Duplicates in text: 0
Null in summary: 0
Duplicates in summary: 0


In [293]:
stats(final_val_data, 'train')

train : (1500, 2)
Null in text: 0
Duplicates in text: 0
Null in summary: 0
Duplicates in summary: 0


In [294]:
stats(final_test_data, 'train')

train : (245, 2)
Null in text: 0
Duplicates in text: 0
Null in summary: 0
Duplicates in summary: 0


<h3>~14 percent train-test split</h3>
<h3>~14 percent val-test split</h3>

In [295]:
final_train_data.to_csv('../for_BERT/uclean_data/train.csv', index = False)
final_test_data.to_csv('../for_BERT/uclean_data/test.csv', index = False)
final_val_data.to_csv('../for_BERT/uclean_data/val.csv', index = False)

### So, finally, 
### the training data has 10520 examples, 
### the validation data has 1500 examples,
### the test data has 245 examples